dashboard/forms/data_group.py

```diff
@@ -89,6 +94,128 @@ def save(self):
+    upc = field_for_model(Product, "upc", required=False)
+    url = field_for_model(Product, "url")
+    brand_name = field_for_model(Product, "brand_name")
+    size = field_for_model(Product, "size")
+    color = field_for_model(Product, "color")
+    item_id = field_for_model(Product, "item_id")
+    parent_item_id = field_for_model(Product, "parent_item_id")
+    short_description = field_for_model(Product, "short_description")
+    long_description = field_for_model(Product, "long_description")
+    thumb_image = field_for_model(Product, "thumb_image")
+    medium_image = field_for_model(Product, "medium_image")
+    large_image = field_for_model(Product, "large_image")
+    model_number = field_for_model(Product, "model_number")
+    manufacturer = field_for_model(Product, "manufacturer")
+
+    def __init__(self, *args, **kwargs):
+        super(ProductCSVForm, self).__init__(*args, **kwargs)
+
+    def clean(self):
+        self.cleaned_data["upc"] = self.cleaned_data.get(
+            "upc", Product.objects.next_upc()
+        )
+        super().clean()
+
+    """
+    ProductBulkCSVFormset
+    Multiple products can be created for a single document. 
+    If user attempts to upload product data for a document 
+    which already has an associated product, a new product 
+    is created with the newly uploaded data (rather than 
+    trying to insert data into the existing product record). 
+    If a user uploads a product file which includes multiple 
+    rows for a single data document ID, each row should be 
+    assumed to be a different product for that document
+    """
+
+
+class ProductBulkCSVFormSet(DGFormSet):
+    prefix = "products"
+    serializer = CSVReader
+
+    def __init__(self, *args, **kwargs):
+        self.form = ProductCSVForm
+        super().__init__(*args, **kwargs)
+
+    def clean(self, *args, **kwargs):
+        header = list(self.bulk.fieldnames)
+        header_cols = [
+            "data_document_id",
+            "data_document_filename",
+            "title",
+            "upc",
+            "url",
+            "brand_name",
+            "size",
+            "color",
+            "item_id",
+            "parent_item_id",
+            "short_description",
+            "long_description",
+            "thumb_image",
+            "medium_image",
+            "large_image",
+            "model_number",
+            "manufacturer",
+        ]
+        if header != header_cols:
+            raise forms.ValidationError(f"CSV column titles should be {header_cols}")
+
+    def save(self):
+        new_products = []
+        new_productdocuments = []
+        rejected_docids = []
+        reports = ""
+        for f in self.forms:
+            f.cleaned_data["created_at"] = datetime.now()
+            product_dict = clean_dict(f.cleaned_data, Product)
+            # Only create the Product if it doesn't already exist
+            if (
+                product_dict["upc"]
```

Comment:
```suggestion
                product_dict.get("upc")
```
There is a possibility that this will otherwise result in a `KeyError`.